dart/common/Addon.h

```diff
@@ -0,0 +1,305 @@
+class AddonWithProtectedState : public Addon
+{
+public:
+
+  using State = Addon::StateMixer<StateData>;
+  constexpr static void (*UpdateState)(Base*) = updateState;
+
+  AddonWithProtectedState() = delete;
+  AddonWithProtectedState(const AddonWithProtectedState&) = delete;
+
+  /// Construct using a StateData instance
+  AddonWithProtectedState(
+      AddonManager* mgr, const StateData& state = StateData());
+
+  // Documentation inherited
+  void setAddonState(
+      const std::unique_ptr<Addon::State>& otherState) override final;
+
+  // Documentation inherited
+  const Addon::State* getAddonState() const override final;
+
+  /// Set the State of this Addon
+  void setState(const StateData& state);
+
+  /// Get the State of this Addon
+  const State& getState() const;
+
+  // Documentation inherited
+  std::unique_ptr<Addon> cloneAddon(
+      AddonManager* newManager) const override final;
+
+protected:
+
+  /// State of this Addon
+  State mState;
+};
+
+//==============================================================================
+/// AddonWithProtectedProperties generates implementations of the Property
+/// managing functions for an Addon class.
+template <class Base, typename PropertiesData,
+          class ManagerType = AddonManager,
+          void (*updateProperties)(Base*) = &detail::NoOp<Base*> >
+class AddonWithProtectedProperties : public Addon
+{
+public:
+
+  using Properties = Addon::PropertiesMixer<PropertiesData>;
+  constexpr static void (*UpdateProperties)(Base*) = updateProperties;
+
+  AddonWithProtectedProperties() = delete;
+  AddonWithProtectedProperties(const AddonWithProtectedProperties&) = delete;
+
+  /// Construct using a PropertiesData instance
+  AddonWithProtectedProperties(
+      AddonManager* mgr, const PropertiesData& properties = PropertiesData());
+
+  // Documentation inherited
+  void setAddonProperties(
+      const std::unique_ptr<Addon::Properties>& someProperties) override final;
+
+  // Documentation inherited
+  const Addon::Properties* getAddonProperties() const override final;
+
+  /// Set the Properties of this Addon
+  void setProperties(const PropertiesData& properties);
+
+  /// Get the Properties of this Addon
+  const Properties& getProperties() const;
+
+  // Documentation inherited
+  std::unique_ptr<Addon> cloneAddon(
+      AddonManager* newManager) const override final;
+
+protected:
+
+  /// Properties of this Addon
+  Properties mProperties;
+};
+
+//==============================================================================
+/// AddonWithProtectedStateAndProperties combines the
+/// AddonWithProtectedState and AddonWithProtectedProperties classes into a
+/// single templated class
+template <class Base, typename StateData, typename PropertiesData,
+          class ManagerType = AddonManager,
+          void (*updateState)(Base*) = &detail::NoOp<Base*>,
+          void (*updateProperties)(Base*) = updateState>
+class AddonWithProtectedStateAndProperties : public Addon
```

Comment:
Yes, I originally tried to implement it as you would expect, but the diamond of death problem was too strong in this case. Normally we could have gotten around the diamond of death by having ``AddonWithProtectedState`` and ``AddonWithProtectedProperties`` both *virtually* inherit ``Addon``, but then it would not be possible to ``static_cast`` from an ``Addon`` to either of those classes (or to any class that inherits them), because the casting would require dynamic dispatch. So avoiding multiple inheritance allows us to avoid doing dynamic dispatch every time we want to grab an ``Addon`` from its manager.